src/datomish/db.cljc

```diff
@@ -12,31 +12,599 @@
+
+  (<apply-datoms
+    [db datoms]
+    "Apply datoms to the store.")
+
+  (<advance-tx
     [db]
-    "Close this database. Returns a pair channel of [nil error]."))
+    "TODO: document this interface."))
+
+(defn db? [x]
+  (and (satisfies? IDB x)
+       (satisfies? IClock x)))
+
+(defn- row->Datom [schema row]
+  (let [e (:e row)
+        a (:a row)
+        v (:v row)]
+    (Datom. e a (ds/<-SQLite schema a v) (:tx row) (and (some? (:added row)) (not= 0 (:added row))))))
 
-(defrecord DB [sqlite-connection]
+(defn- <insert-fulltext-value [db value]
+  (go-pair
+    ;; This dance is necessary to keep fulltext_values.text unique.  We want uniqueness so that we
+    ;; can work with string values, maintaining consistency throughout the transactor
+    ;; implementation.  (Without this, we'd need to handle a [rowid text] pair specially when
+    ;; comparing in the transactor.)  Unfortunately, it's not possible to declare a unique
+    ;; constraint on a virtual table, including an FTS table.  External content tables (see
+    ;; http://www.sqlite.org/fts3.html#section_6_2_2) don't appear to address our use case, so we
+    ;; maintain uniqueness ourselves.
+    (let [rowid
+          (if-let [row (first (<? (s/all-rows (:sqlite-connection db) ["SELECT rowid FROM fulltext_values WHERE text = ?" value])))]
+            (:rowid row)
+            (do
+              (<? (s/execute! (:sqlite-connection db) ["INSERT INTO fulltext_values VALUES (?)" value]))
+              (:rowid (first (<? (s/all-rows (:sqlite-connection db) ["SELECT last_insert_rowid() AS rowid"]))))))
+          ]
+      rowid)))
+
+(defrecord DB [sqlite-connection schema idents current-tx]
+  ;; idents is map {ident -> entid} of known idents.  See http://docs.datomic.com/identity.html#idents.
   IDB
   (query-context [db] (context/->Context (source/datoms-source db) nil nil))
-  (close [db] (s/close (.-sqlite-connection db))))
 
-(defn <with-sqlite-connection [sqlite-connection]
+  (schema [db] (.-schema db))
+
+  (idents [db] (.-idents db))
+
+  (current-tx
+    [db]
+    (inc (:current-tx db)))
+
+  ;; TODO: use q for searching?  Have q use this for searching for a single pattern?
+  (<eavt [db pattern]
+    (let [[e a v] pattern
+          v (and v (ds/->SQLite schema a v))] ;; We assume e and a are always given.
+      (go-pair
+        (->>
+          {:select [:e :a :v :tx [1 :added]] ;; TODO: generalize columns.
+           :from [:all_datoms]
+           :where (cons :and (map #(vector := %1 %2) [:e :a :v] (take-while (comp not nil?) [e a v])))} ;; Must drop nils.
+          (sql/format)
+
+          (s/all-rows (:sqlite-connection db))
+          (<?)
+
+          (mapv (partial row->Datom (.-schema db))))))) ;; TODO: understand why (schema db) fails.
+
+  (<avet [db pattern]
+    (let [[a v] pattern
+          v (ds/->SQLite schema a v)]
+      (go-pair
+        (->>
+          {:select [:e :a :v :tx [1 :added]] ;; TODO: generalize columns.
```

Comment:
You can format this query (and the other fixed queries above) once, rather than once per execution. That'll save a lot of runtime.